default.nix

```diff
@@ -1,22 +1,92 @@
-  callPackage = pkgs'.python27Packages.callPackage;
+  sources = import nix/sources.nix;
 in
-callPackage ./zkapauthorizer.nix {
-  challenge-bypass-ristretto = callPackage ./python-challenge-bypass-ristretto.nix { };
-  inherit hypothesisProfile collectCoverage testSuite trialArgs;
-  tahoe-lafs = tahoe-lafs';
-}
+{ pkgs ? import sources.release2015 {}
+, pypiData ? sources.pypi-deps-db
+, mach-nix ? import sources.mach-nix { inherit pkgs pypiData; }
+, tahoe-lafs-source ? "tahoe-lafs"
+, tahoe-lafs-repo ? sources.${tahoe-lafs-source}
+}:
+  let
+    lib = pkgs.lib;
+    python = "python27";
+    providers = {
+      _default = "sdist,nixpkgs,wheel";
+      # mach-nix doesn't provide a good way to depend on mach-nix packages,
+      # so we get it as a nixpkgs dependency from an overlay. See below for
+      # details.
+      tahoe-lafs = "nixpkgs";
+      # not packaged in nixpkgs at all, we can use the binary wheel from
+      # pypi though.
+      python-challenge-bypass-ristretto = "wheel";
+      # Pure python packages that don't build correctly from sdists
+      # - patches in nixpkgs that don't apply
+      boltons = "wheel";
+      chardet = "wheel";
+      urllib3 = "wheel";
+      # - incorrectly detected dependencies due to pbr
+      fixtures = "wheel";
+      testtools = "wheel";
+      traceback2 = "wheel";
+      # - Incorrectly merged extras - https://github.com/DavHau/mach-nix/pull/334
+      tqdm = "wheel";
+    };
+  in
+    rec {
+      tahoe-lafs = mach-nix.buildPythonPackage rec {
+        inherit python providers;
+        name = "tahoe-lafs";
+        version = "1.16.post999";
+        # See https://github.com/DavHau/mach-nix/issues/190
+        requirementsExtra = ''
+          pyrsistent < 0.17
+          foolscap == 0.13.1
+          configparser
+          eliot
+        '';
+        postPatch = ''
+          cat > src/allmydata/_version.py <<EOF
+          # This _version.py is generated by nix.
+
+          verstr = "${version}+git-${tahoe-lafs-repo.rev}"
+          __version__ = verstr
+          EOF
+        '';
+        src = tahoe-lafs-repo;
+      };
+      zkapauthorizer = mach-nix.buildPythonApplication rec {
+        inherit python providers;
+        src = ./.;
+        # mach-nix does not provide a way to specify dependencies on other
+        # mach-nix packages, that incorporates the requirements and overlays
+        # of that package.
+        # See https://github.com/DavHau/mach-nix/issues/123
+        # In particular, we explicitly include the requirements of tahoe-lafs
+        # here, and include it in a python package overlay.
+        requirementsExtra = tahoe-lafs.requirements;
+        overridesPre = [
+          (
+            self: super: {
+              inherit tahoe-lafs;
+            }
+          )
+        ];
+        # Record some settings here, so downstream nix files can consume them.
+        meta.mach-nix = { inherit python providers; };
+      };
+
+      privatestorage = let
```

Comment:
I added this derivation, and turned `default.nix` into an attrset rather than a derivation to get the right binaries into the derivation used by PrivateStorageio and also expose the `zkaupauthorizer` derivation to `tests.nix`. It'd also be possible to expose the later as an attribute on this derivation; I'm not sure which approach is better (if either).